scripts/version-utils.js

```diff
@@ -9,28 +9,147 @@
+ * ```
+ * {
+ *   version: string,
+ *   major: number,
+ *   minor: number,
+ *   patch: number,
+ *   prerelease: string
+ * }
+ * ```
+ */
+function parseVersion(versionStr, buildType) {
+  validateBuildType(buildType);
+
+  const match = extractMatchIfValid(versionStr);
   const [, version, major, minor, patch, prerelease] = match;
-  return {
+
+  const versionObject = {
     version,
     major,
     minor,
     patch,
     prerelease,
   };
+
+  validateVersion(versionObject, buildType);
+
+  return versionObject;
+}
+
+function validateBuildType(buildType) {
+  const validBuildTypes = new Set(['release', 'dry-run', 'nightly']);
+  if (!validBuildTypes.has(buildType)) {
+    throw new Error(`Unsupported build type: ${buildType}`);
+  }
+}
+
+function extractMatchIfValid(versionStr) {
+  const match = versionStr.match(VERSION_REGEX);
+  if (!match) {
+    throw new Error(
+      `You must pass a correctly formatted version; couldn't parse ${versionStr}`,
+    );
+  }
+  return match;
+}
+
+function validateVersion(versionObject, buildType) {
+  const map = {
+    release: validateRelease,
+    'dry-run': validateDryRun,
+    nightly: validateNightly,
+  };
+
+  const validationFunction = map[buildType];
+  validationFunction(versionObject);
+}
+
+/**
+ * Releases are in the form of 0.Y.Z[-RC.0]
+ */
+function validateRelease(version) {
+  const validRelease = isStableRelease(version) || isStablePrerelease(version);
+  if (!validRelease) {
+    throw new Error(`Version ${version.version} is not valid for Release`);
+  }
+}
+
+function validateDryRun(version) {
+  const isNightly = is000(version) && version.prerelease != null;
+
+  if (
+    !isMain(version) &&
+    !is1000_0_1(version) &&
+    !isNightly &&
+    !isStableRelease(version) &&
+    !isStablePrerelease(version)
+  ) {
+    throw new Error(`Version ${version.version} is not valid for dry-runs`);
+  }
+}
+
+function validateNightly(version) {
+  // a valid nightly is a prerelease
+  const isPrerelease = version.prerelease != null;
+  const isValidNightly = is000(version) && isPrerelease;
+  if (!isValidNightly) {
+    throw new Error(`Version ${version.version} is not valid for nightlies`);
+  }
+}
+
+function isStableRelease(version) {
+  return (
+    version.major === '0' && version.minor !== '0' && version.prerelease == null
+  );
+}
+
+function isStablePrerelease(version) {
+  return (
+    version.major === '0' &&
+    version.minor !== '0' &&
+    version.prerelease != null &&
+    (version.prerelease.startsWith('rc.') ||
+      version.prerelease.startsWith('rc-'))
+  );
+}
+
+function is000(version) {
+  return (
+    version.major === '0' && version.minor === '0' && version.patch === '0'
+  );
+}
+
+function isMain(version) {
+  return (
+    version.major === '1000' && version.minor === '0' && version.patch === '0'
+  );
+}
+
+function is1000_0_1(version) {
+  return (
+    version.major === '1000' && version.minor === '0' && version.patch === '1'
+  );
```

Comment:
what is this case covering? 🤔

Reply:
this: https://github.com/facebook/react-native/blob/main/.circleci/config.yml#L1646

I think we can change this, but I left it here to highlight this and let you and Nico spread some knowledge about why this value has been chosen! :D

Reply:
that was not taken by either of us, looks like it's coming from here https://github.com/facebook/react-native/commit/e4b5d3eec9bae44c46795c4be097b31cb48593ae (and I was very much no aware of that)

soooo yeah. I guess we need to keep it around for now? It looks like it's just doing a dry run, so maybe realign it with the versioning for the dry runs? Also, the more I look at that PR, the more it feels wrong that the `package_and_publish_release_dryrun` is using a fixed version and not the one set in the branch :/

might be worth a refactoring of that part of circleci

Reply:
Ok, I tried to pick up the version from the package.json for this problem, but it can't work. The reason is that the `set-rn-version` script fails if we try to set the same version that is already in the `package.json`. That's why they used the `v1000.0.1`, because they were sure that it would have worked.

We can add a bit of extra logic to dry-run on a patch: this will work for dry-running on `main` and on `stable` branches. However, it won't be exact for RCs. However, for RCs, it is very hard to properly bump them: the `prerelease` part of the version is handled as a string (because sometimes it is a date - for the nightlies -, some other times it is the RC.x, and in the test local is something different again), so we can't actually bump the patch.

as an approximation, given that it is just a dry-run and given that we always run it with 1000.0.1, what if we accept that for RCs we dry run on a schema that is 0.X.Y-RC.K?
@kelset @cortinico